File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.py
import sys, math
input = sys.stdin.readline
from collections import deque

N = int(input())
arr = deque([int(i) for i in input().split()])

result = deque([])
place = deque([])

def dokidoki():
    cnt = 1
    while arr:

        if place:
            if place[-1] == cnt:
                result.append(place.pop())
                cnt += 1
                continue

        if arr:
            x = arr.popleft()
            if cnt == x:
                result.append(x)
                cnt += 1
            else:
                place.append(x)

    while place:

        if place[-1] == cnt:
            result.append(place.pop())
            cnt += 1
        else:
            return "Sad"

    return "Nice"

print(dokidoki())

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.js
let fs = require('fs')
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n")
// let input = fs.readFileSync("./EX.txt").toString().trim().split('\n')

let N = parseInt(input[0])
let arr = input[1].split(" ").map(x => parseInt(x))

stack = []

cnt = 1
while (arr.length) {
    x = arr.shift()
    stack.push(x)

    while (stack.length && stack[stack.length-1] == cnt){
        stack.pop()
        cnt += 1
    }
}

if (stack.length) {
    console.log("Sad")
} else {
    console.log("Nice")
}